src/CodeGen_WebAssembly.cpp

```diff
@@ -26,94 +26,67 @@ CodeGen_WebAssembly::CodeGen_WebAssembly(Target t)
-        // N.B. Saturating subtracts are expressed by widening to a *signed* type
-        {Target::WasmSimd128, true, Int(8, 16), 0, "llvm.wasm.sub.saturate.signed.v16i8", i8_sat(wild_i16x_ - wild_i16x_)},
-        {Target::WasmSimd128, true, UInt(8, 16), 0, "llvm.wasm.sub.saturate.unsigned.v16i8", u8_sat(wild_i16x_ - wild_i16x_)},
-        {Target::WasmSimd128, true, Int(16, 8), 0, "llvm.wasm.sub.saturate.signed.v8i16", i16_sat(wild_i32x_ - wild_i32x_)},
-        {Target::WasmSimd128, true, UInt(16, 8), 0, "llvm.wasm.sub.saturate.unsigned.v8i16", u16_sat(wild_i32x_ - wild_i32x_)},
-
-        //{Target::WasmSimd128, true, UInt(8, 16), 0, "llvm.wasm.avgr.unsigned.v16i8", u8(((wild_u16x_ + wild_u16x_) + 1) / 2)},
-        //{Target::WasmSimd128, true, UInt(8, 16), 0, "llvm.wasm.avgr.unsigned.v16i8", u8(((wild_u16x_ + wild_u16x_) + 1) >> 1)},
-        //{Target::WasmSimd128, true, UInt(16, 8), 0, "llvm.wasm.avgr.unsigned.v8i16", u16(((wild_u32x_ + wild_u32x_) + 1) / 2)},
-        //{Target::WasmSimd128, true, UInt(16, 8), 0, "llvm.wasm.avgr.unsigned.v8i16", u16(((wild_u32x_ + wild_u32x_) + 1) >> 1)},
-
-        // TODO: LLVM should support this directly, but doesn't yet.
-        // To make this work, we need to be able to call the intrinsics with two vecs.
-        // @abadams sez: "The way I've had to do this in the past is with force-inlined implementations
-        // that accept the wider vec, e.g. see packsswbx16 in src/runtime/x86.ll"
-        // {Target::WasmSimd128, false, Int(8, 16), 0, "llvm.wasm.narrow.signed.v16i8.v8i16", i8(wild_i16x_)},
-        // {Target::WasmSimd128, false, Int(16, 8), 0, "llvm.wasm.narrow.signed.v8i16.v4i32", i16(wild_i32x_)},
-        // {Target::WasmSimd128, false, UInt(8, 16), 0, "llvm.wasm.narrow.unsigned.v16i8.v8i16", u8(wild_u16x_)},
-        // {Target::WasmSimd128, false, UInt(16, 8), 0, "llvm.wasm.narrow.unsigned.v8i16.v4i32", u16(wild_u32x_)},
-    };
-
-    for (size_t i = 0; i < sizeof(patterns) / sizeof(patterns[0]); i++) {
-        const Pattern &pattern = patterns[i];
-
-        if (!target.has_feature(pattern.feature)) {
+namespace {
+
+constexpr int max_intrinsic_args = 4;
+
+struct WasmIntrinsic {
+    const char *intrin_name;
+    halide_type_t ret_type;
+    const char *name;
+    halide_type_t arg_types[max_intrinsic_args];
+    Target::Feature feature = Target::FeatureEnd;
+};
+
+// clang-format off
+const WasmIntrinsic intrinsic_defs[] = {
+    {"llvm.sadd.sat.v8i16", Int(16, 8), "saturating_add", {Int(16, 8), Int(16, 8)}, Target::WasmSimd128},
+    {"llvm.uadd.sat.v8i16", UInt(16, 8), "saturating_add", {UInt(16, 8), UInt(16, 8)}, Target::WasmSimd128},
+    {"llvm.sadd.sat.v16i8", Int(8, 16), "saturating_add", {Int(8, 16), Int(8, 16)}, Target::WasmSimd128},
+    {"llvm.uadd.sat.v16i8", UInt(8, 16), "saturating_add", {UInt(8, 16), UInt(8, 16)}, Target::WasmSimd128},
+
+    // TODO: Are these really different than the standard llvm.*sub.sat.*?
```

Comment:
When I first wrote this code, they were necessary in order to get the right instructions emitted. Maybe trunk LLVM has improved since then.